openwrt-mdnsd-cve-2020-11750/src/fuzzer.c

```diff
@@ -0,0 +1,369 @@
+#endif
+    }
+
+    DBG(1, "Q -> %s %s\n", dns_type_string(q->type), name);
+
+    switch (q->type) {
+    case TYPE_ANY:
+        if (!strcmp(name, mdns_hostname_local)) {
+#if 0
+            dns_reply_a(iface, to, announce_ttl);
+            service_reply(iface, to, NULL, NULL, announce_ttl);
+#endif
+        }
+        break;
+
+    case TYPE_PTR:
+        if (!strcmp(name, C_DNS_SD)) {
+#if 0
+            dns_reply_a(iface, to, announce_ttl);
+            service_announce_services(iface, to, announce_ttl);
+#endif
+        } else {
+            if (name[0] == '_') {
+#if 0
+                service_reply(iface, to, NULL, name, announce_ttl);
+#endif
+            } else {
+                /* First dot separates instance name from the rest */
+                char *dot = strchr(name, '.');
+
+                if (dot) {
+                    *dot = '\0';
+#if 0
+                    service_reply(iface, to, name, dot + 1, announce_ttl);
+#endif
+                    *dot = '.';
+                }
+            }
+        }
+        break;
+
+    case TYPE_AAAA:
+    case TYPE_A:
+        host = strstr(name, ".local");
+        if (host)
+            *host = '\0';
+        if (!strcmp(umdns_host_label, name))
+        {
+#if 0
+            dns_reply_a(iface, to, announce_ttl);
+#endif
+        }
+        break;
+    };
+}
+
+void
+dns_handle_packet(struct interface *iface, struct sockaddr *from, uint16_t port, uint8_t *buffer, int len)
+{
+    struct dns_header *h;
+    uint8_t *b = buffer;
+    int rlen = len;
+
+    h = dns_consume_header(&b, &rlen);
+    if (!h) {
+        fprintf(stderr, "dropping: bad header\n");
+        return;
+    }
+
+#if 0
+    if (h->questions && !iface->multicast && port != MCAST_PORT)
+        /* silently drop unicast questions that dont originate from port 5353 */
+        return;
+#endif
+
+    while (h->questions-- > 0) {
+        char *name = dns_consume_name(buffer, len, &b, &rlen);
+        struct dns_question *q;
+
+        if (!name) {
+            fprintf(stderr, "dropping: bad name\n");
+            return;
+        }
+
+        q = dns_consume_question(&b, &rlen);
+        if (!q) {
+            fprintf(stderr, "dropping: bad question\n");
+            return;
+        }
+
+        if (!(h->flags & FLAG_RESPONSE))
+            parse_question(iface, from, name, q);
+    }
+
+    if (!(h->flags & FLAG_RESPONSE))
+        return;
+
+    while (h->answers-- > 0)
+        if (parse_answer(iface, from, buffer, len, &b, &rlen, 1))
+            return;
+
+    while (h->authority-- > 0)
+        if (parse_answer(iface, from, buffer, len, &b, &rlen, 1))
+            return;
+
+    while (h->additional-- > 0)
+        if (parse_answer(iface, from, buffer, len, &b, &rlen, 1))
+            return;
+
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+    uint8_t* buffer = malloc(size);
+    memcpy(buffer, data, size);
+    dns_handle_packet(NULL, NULL, 1234, buffer, size);
```

Comment:
Question: do you really need the memcpy?  Why not pass `data` directly?

Reply:
To prevent:

```
==21925== ERROR: libFuzzer: fuzz target overwrites it's const input
```

so I copy the input to a buffer that can be modified.